bin/cron/pardot_import

```diff
@@ -197,19 +188,189 @@ def main
 
     # As a sniff test, build the URL that would result from our current prospect list
     # so we can see how long it is
-    url = build_create_batch_prospects_url(prospects)
+    url = build_batch_prospects_url(prospects, config)
 
     # If the URL is longer than an empirically determined maximum safe length, or if
     # we have hit our max # of prospects allowed by Pardot in one batch API call, go ahead
     # and submit the batch
     if url.length > URL_LENGTH_SEND_THRESHOLD || prospects.size == MAX_PROSPECT_BATCH_SIZE
-      submit_create_batch prospects
+      submit_prospect_batch(prospects, config)
+      num_operations += prospects.length
+      if num_operations > num_operations_last_print + 1000
+        log "Prospect total operations so far: #{num_operations}"
+        num_operations_last_print = num_operations
+      end
       prospects = []
     end
   end
 
-  # submit any batch remainder
-  submit_create_batch prospects unless prospects.empty?
+  # Submit any batch remainder
+  submit_prospect_batch(prospects, config) unless prospects.empty?
+  num_operations += prospects.length
+  log "Contact #{config[:operation_name]} pass completed. #{num_operations} total operations."
+end
+
+# Create or update a batch of prospects
+# @param prospects [Array<Hash>] array of hashes of prospect data
+# @param config [Hash] hash of config params to use to control create vs update behavior
+# @returns [Nokogiri::XML] XML response from Pardot
+def submit_prospect_batch(prospects, config)
+  # Build the URL containing prospect data. (Prospect data is sent as a JSON blob in a
+  # query parameter)
+  url = build_batch_prospects_url(prospects, config)
+
+  # Build array of the emails for to create in Pardot
+  prospect_emails = prospects.collect{ |x| x["email"] }.to_a
+  malformed_emails = []
+
+  num_submitted = prospects.length
+
+  # Post the data to create or update a batch of prospects
+  doc = post_with_auth_retry(url)
+  status = doc.xpath('/rsp/@stat').text
+
+  # Look for errors within the batch and handle them individually.
+  if status != "ok"
+    num_removed = 0
+    doc.xpath('/rsp/errors/*').each do |node|
+      error_text = node.text
+      if error_text.eql?("Invalid prospect email address")
+        # Identifier of contact with error is zero-based index
+        prospect_identifier = node.attr("identifier").to_i
+        # Adjust contact index for any we have already removed in this batch
+        prospect_identifier -= num_removed
+
+        malformed_email = prospect_emails[prospect_identifier]
+        log "Pardot reported email \"#{malformed_email}\" as malformed, marking as malformed in our DB"
+        # Add the rejected email to our list of malformed emails
+        malformed_emails << malformed_email
+        prospect_emails.delete_at(prospect_identifier)
```

Comment:
You can eliminate a line if you take advantage of the fact that `delete_at` returns the deleted element.
https://docs.ruby-lang.org/en/2.0.0/Array.html#method-i-delete_at
```
malformed_email = prospect_emails.delete_at(prospect_identifier)
log "Pardot reported email \"#{malformed_email}\" as malformed, marking as malformed in our DB"
malformed_emails << malformed_email
```